dialogue-core/src/main/java/com/palantir/dialogue/core/LeakDetectingChannel.java

```diff
@@ -0,0 +1,187 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.collect.ListMultimap;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Optional;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class LeakDetectingChannel implements Channel {
+
+    private static final Logger log = LoggerFactory.getLogger(LeakDetectingChannel.class);
+
+    private final Channel delegate;
+    private final String channelName;
+    private final DialogueClientMetrics metrics;
+
+    LeakDetectingChannel(Channel delegate, String channelName, DialogueClientMetrics metrics) {
+        this.delegate = delegate;
+        this.channelName = channelName;
+        this.metrics = metrics;
+    }
+
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        return Futures.transform(
+                delegate.execute(endpoint, request),
+                response -> new LeakDetectingResponse(response, new LeakDetector(response, endpoint)),
+                MoreExecutors.directExecutor());
+    }
+
+    /**
+     * {@link LeakDetector} object is shared between the {@link Response} and {@link Response#body()} to ensure
+     * at least one of the two has been closed.
+     */
+    private final class LeakDetector {
+
+        private final Endpoint endpoint;
+        private final Response response;
+
+        @Nullable
+        private final Throwable creationTrace;
+
+        private boolean armed = true;
+
+        LeakDetector(Response response, Endpoint endpoint) {
+            this.response = response;
+            this.endpoint = endpoint;
+            this.creationTrace = log.isTraceEnabled() ? new SafeRuntimeException("created here") : null;
+        }
+
+        void disarm() {
+            armed = false;
+        }
+
+        @Override
+        @SuppressWarnings("NoFinalizer")
+        protected void finalize() throws Throwable {
```

Comment:
Seems like the finalize method is deprecated in java9 with a recommendation to use `java.lang.ref.Cleaner` or PhantomReference instead - https://bugs.openjdk.java.net/browse/JDK-8165641... seems like it should just be a matter of calling cleaner.register with a nice runnable? https://docs.oracle.com/javase/9/docs/api/java/lang/ref/Cleaner.html

Reply:
We can’t do that with jdk8 though.

Reply:
Ah dang ok

Reply:
If this method ever manages to block then do we end up preventing all garbage collection from succeeding?

Reply:
It can block the finalizer queue from progressing, but that should only occur if we have leaked a resource. I’d prefer to have the data to resolve the situation in that scenario though.

Reply:
yep I think that makes sense!